### PR TITLE
Concurrency for dictionary generation

### PR DESCRIPTION
**Pull request by ashokblend**
1.Change lock mechanism to lock on any files
2.Change dictionary generation code to handle concurrency

### CODE REVIEW

File: core/src/main/java/org/carbondata/core/writer/sortindex/CarbonDictionarySortIndexWriterImpl.java [hunk 1]
```diff
@@ -146,7 +149,39 @@ private void writeColumnSortInfo() throws IOException {
...
+        carbonTablePath.getSortIndexFilePath(columnIdentifier.getColumnId(), dictOffset);
+    cleanUpOldSortIndex(carbonTablePath);
+  }
+
+  /**
+   * It cleans up old unused sortindex file
+   * @param carbonTablePath
+   */
+  private void cleanUpOldSortIndex(CarbonTablePath carbonTablePath) {
+    CarbonFile[] files = carbonTablePath
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/core/writer/sortindex/CarbonDictionarySortIndexWriterImpl.java]: always exempt last file

File: core/src/main/java/org/carbondata/core/writer/sortindex/CarbonDictionarySortIndexWriterImpl.java [hunk 2]
```diff
@@ -146,7 +149,39 @@ private void writeColumnSortInfo() throws IOException {
...
+      maxTime = CarbonCommonConstants.DEFAULT_MAX_QUERY_EXECUTION_TIME;
+    }
+    if (null != files) {
+      for (CarbonFile carbonFile : files) {
+        long difference = System.currentTimeMillis() - carbonFile.getLastModifiedTime();
+        long minutesElapsed = (difference / (1000 * 60));
+        if (minutesElapsed > maxTime) {
+          if (!carbonFile.delete()) {
+            LOGGER.warn("Failed to delete sortindex file." + carbonFile.getAbsolutePath());
+          }
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/core/writer/sortindex/CarbonDictionarySortIndexWriterImpl.java]: delete success also please log

File: integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala [hunk 3]
```diff
@@ -260,36 +259,57 @@ class CarbonGlobalDictionaryGenerateRDD(
...
+            )
+          } else {
+            null
+          }
+          val dictCacheTime = (System.currentTimeMillis - t2)
+          val dictWriteTask = new DictionaryWriterTask(valuesBuffer,
+            dictionaryForDistinctValueLookUp,
+            model,
+            split.index)
+          val sortIndexWriteTask = new SortIndexWriterTask(model,
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala]: execute task and later fetch required result from task and pass it to next task.

File: integration/spark/src/main/scala/org/carbondata/spark/tasks/SortIndexWriterTask.scala [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        preparator.getDictionarySortInfo(dictStats.distinctValues, dictionary,
+          model.primDimensions(index).getDataType)
+      carbonDictionarySortIndexWriter =
+        dictService.getDictionarySortIndexWriter(model.table, model.columnIdentifier(index),
+          model.hdfsLocation)
+      carbonDictionarySortIndexWriter.writeSortIndex(dictionarySortInfo.getSortIndex)
+      carbonDictionarySortIndexWriter
+        .writeInvertedSortIndex(dictionarySortInfo.getSortIndexInverted)
+    }
+    DictionaryStats(dictStats.distinctValues,
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/carbondata/spark/tasks/SortIndexWriterTask.scala]: make sure that all close moved to finally with a null check

File: core/src/main/java/org/carbondata/core/writer/sortindex/CarbonDictionarySortIndexWriterImpl.java (hunk 1 above)
Comment (ashokblend) [inline core/src/main/java/org/carbondata/core/writer/sortindex/CarbonDictionarySortIndexWriterImpl.java]: It's handled please check cleanUpOldSortIndex method

File: core/src/main/java/org/carbondata/core/writer/sortindex/CarbonDictionarySortIndexWriterImpl.java (hunk 2 above)
Comment (ashokblend) [inline core/src/main/java/org/carbondata/core/writer/sortindex/CarbonDictionarySortIndexWriterImpl.java]: Fixed

File: integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala (hunk 3 above)
Comment (ashokblend) [inline integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala]: It's handled

File: integration/spark/src/main/scala/org/carbondata/spark/tasks/SortIndexWriterTask.scala (hunk 4 above)
Comment (ashokblend) [inline integration/spark/src/main/scala/org/carbondata/spark/tasks/SortIndexWriterTask.scala]: It's handled

File: core/src/main/java/org/carbondata/core/writer/sortindex/CarbonDictionarySortIndexWriterImpl.java [hunk 5]
```diff
@@ -172,12 +178,19 @@ private void cleanUpOldSortIndex(CarbonTablePath carbonTablePath) {
...
+      for (int i = 0; i < files.length - 1; i++) {
+        long difference = System.currentTimeMillis() - files[i].getLastModifiedTime();
         long minutesElapsed = (difference / (1000 * 60));
         if (minutesElapsed > maxTime) {
-          if (!carbonFile.delete()) {
-            LOGGER.warn("Failed to delete sortindex file." + carbonFile.getAbsolutePath());
+          if (!files[i].delete()) {
+            LOGGER.warn("Failed to delete sortindex file." + files[i].getAbsolutePath());
+          } else {
+            LOGGER.info("Sort index file is deleted." + files[i].getAbsolutePath());
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/core/writer/sortindex/CarbonDictionarySortIndexWriterImpl.java]: After deletion can we get absolute path?

File: core/src/test/java/org/carbondata/core/writer/CarbonDictionaryWriterImplTest.java [hunk 6]
```diff
@@ -245,9 +245,15 @@ private CarbonDictionaryWriterImpl prepareWriter() throws IOException {
   private void overwriteDictionaryMetaFile(ColumnDictionaryChunkMeta firstDictionaryChunkMeta,
       String dictionaryFile) throws IOException {
     ThriftWriter thriftMetaChunkWriter = new ThriftWriter(dictionaryFile, false);
-    thriftMetaChunkWriter.open();
```
Comment (gvramana) [inline core/src/test/java/org/carbondata/core/writer/CarbonDictionaryWriterImplTest.java]: Add tests to test old sort index files deletion scenario.

**Comment by gvramana** [comment]
rebase code

File: integration/spark/src/test/scala/org/carbondata/spark/util/GlobalDictionaryUtilConcurrentTestCase.scala [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    val carbonTablePath = PathFactory.getInstance()
+        .getCarbonTablePath(columnIdentifier, sampleRelation.cubeMeta.storePath, carbonTableIdentifier);
+    val dictPath = carbonTablePath.getDictionaryFilePath(columnIdentifier.getColumnId)
+    val dictFile = FileFactory.getCarbonFile(dictPath, FileFactory.getFileType(dictPath))
+    val offSet = dictFile.getSize
+    val sortIndexPath = carbonTablePath.getSortIndexFilePath(columnIdentifier.getColumnId, offSet)
+    val sortIndexFile = FileFactory.getCarbonFile(sortIndexPath, FileFactory.getFileType(sortIndexPath))
+    assert(sortIndexFile.exists())
+    val sortIndexFiles = carbonTablePath.getSortIndexFiles(sortIndexFile.getParentFile, columnIdentifier.getColumnId)
+    assert(sortIndexFiles.length == 2)
```
Comment (ashokblend) [inline integration/spark/src/test/scala/org/carbondata/spark/util/GlobalDictionaryUtilConcurrentTestCase.scala]: @gvramana I have added delete test case here.
 I set "-1" as query max time. With this case there will always be 2 sortindex file.
Please check.

**Pull request merged**